src/main/kotlin/com/exactpro/th2/conn/dirty/tcp/core/ActionStreamExecutor.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2021-2021 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.conn.dirty.tcp.core
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Executor
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Executor which ensures that actions of a single stream are executed in their submission order.
+ *
+ * Actions of different streams can be executed concurrently
+ *
+ * @param executor executor to run actions on
+ * @param onError callback for failed executions
+ */
+class ActionStreamExecutor(
+    private val executor: Executor,
+    private val onError: (stream: String, error: Throwable) -> Unit
+) {
+    private val streams = ConcurrentHashMap<String, Action>()
+
+    /**
+     * Schedules execution of [action] in the specified [stream].
+     *
+     * Action will be scheduled for execution immediately if the previous action has been already executed.
+     *
+     * Otherwise, it will be linked to the previous action which schedule it for execution after its completion.
+     *
+     * @param stream action stream
+     * @param action action to execute in the stream
+     */
+    fun execute(stream: String, action: Runnable) {
+        // add backpressure?
```

Comment:
It is a good idea because the box can produce actions faster than the executer complete them. In this case, the application will be late with a reaction to a current situation. Please attention when working with backpressure pauses in Netty, some of the threads should not be blocked.